stackrox/stackrox
# ROX-9896: Refactor authenticated routes to solve integration test failures in UI

## Description


Solution to avoid integration test failures that supersedes work around #1159

1. System Health Integrations local deployment should go to Notifier Integrations anchor on Integrations page via click View All

    CypressError: Timed out retrying after 5000ms: `cy.wait()` timed out waiting `5000ms` for the 1st request to the route: `GetNotifiers`. No request ever occurred.
    ![cypress-screenshot-1](https://user-images.githubusercontent.com/11862657/161602250-b2cc0882-3c2a-4bfd-a26e-774d7e21dfe2.png)

    2022-03-31 https://app.circleci.com/pipelines/github/stackrox/stackrox/8491/workflows/6805edd9-7b40-406f-881e-d758aa6849dc/jobs/375020

    diagnostic-bundle/logimbue.json
    * `"featureFlags": { "featureFlags": [] },`
    * `"Error: Feature Flag (ROX_VULN_REPORTING) is not a valid feature flag"`
    * `"Error: Feature Flag (ROX_SYSTEM_HEALTH_PF) is not a valid feature flag"`

2. Vuln Management Dashboard Page clicking the "Top Riskiest Images" widget's "View All" button should take you to the images list

    AssertionError: Timed out retrying after 4000ms: Expected to find element: `[data-testid="widget"]:contains('Top Riskiest Images')`, but never found it.
    ![cypress-screenshot-2](https://user-images.githubusercontent.com/11862657/161602350-fc84ee98-1608-493e-b3f4-03122f1e0cd9.png)

    2022-04-04 https://app.circleci.com/pipelines/github/stackrox/stackrox/8655/workflows/53f69365-7bae-4df7-b5a1-6a991b35b580/jobs/382789

    diagnostic-bundle/logimbue.json
    * `"featureFlags": { "featureFlags": [] },`
    * `"Error: Feature Flag (ROX_VULN_REPORTING) is not a valid feature flag"`

### Problem

At least in integration testing environment, cannot assume that responses to /v1/featureflags and /v1/mypermissions requests are available:
* src/Containers/MainPage/Body.tsx (see Problem 1 above)
* src/Containers/Navigation/NavigationSideBar/NavigationSideBar.tsx (see Problem 2 above)

In theory, `AppWrapper` element has similiar problem with /v1/config/public request, however:
* timing problem would only affect its tests in systemconfig.test.js
* would not throw in production because it renders with optional chaining

### Solution

Increase cohesion and reduce coupling, especially rendering `ProtectedRoute` at two levels

1. AppPage: render authenticated versus non-authenticated routes
2. AuthenticatedRoutes: render according to authStatus
3. MainPage: render page elements after feature flags and role permissions are available
4. Body: render individual (authenticated) routes with conditional rendering for feature flags, role permissions, and dev mode

Changed files

1. Edit src/Containers/AppPage.js and rename as src/Containers/AppPage.tsx
    * Replace `<ProtectedRoute path={mainPath} component={MainPage} />` with `<Route path={mainPath} component={AuthenticatedRoutes} />`

    https://reactrouter.com/docs/en/v6/upgrading/v5#refactor-custom-routes

    > Replace any elements inside a `<Switch>` that are not plain `<Route>` elements with a regular `<Route>`. This includes any `<PrivateRoute>`-style custom components.

2. Replace src/Components/ProtectedRoute.js with src/Containers/MainPage/AuthenticatedRoutes.tsx
    * Replace `connect` for `authStatus` with `useSelector`
    * Replace implicit `location` prop from `Route` owner with `useLocation`

3. Edit src/Containers/MainPage/Body.tsx
    * Replace `ProtectedRoute` with `Route` (ditto above quote from React Router docs)
    * Replace `ProtectedRoute` props `devOnly` `featureFlagEnabled` `requiredPermission` with conditional rendering
    * Replace `useFeatureFlagEnabled` hook with `isFeatureFlagEnabled` prop

4. Edit src/Containers/MainPage/MainPage.tsx
    * Replace `connect` and props which confused TypeScript when AuthenticatedRoutes.tsx renders `<MainPage />` element without props
    * Replace `mapStateToProps` with `useSelector`
    * Replace `mapDispatchToProps` with `useDispatch` (see Residue 3)
    * Replace `history` prop with `useHistory` (see Residue 3)
    * Add `const { isFeatureFlagEnabled, isLoadingFeatureFlags } = useFeatureFlags();`
    * Add `const { hasReadAccess, isLoadingPermissions } = usePermissions();`
    * Render `LoadingSection` if `isLoadingFeatureFlags` or `isLoadingPermissions` because either `Body` or `NavigationSideBar` might have conditional rendering

5. src/Containers/Navigation/NavigationSideBar/NavigationSideBar.tsx
    * Replace `usePermissions` hook with `hasReadAccess` prop
    * Replace `useFeatureFlagEnabled` hook with `isFeatureFlagEnabled` prop

6. ui/apps/platform/src/hooks/useFeatureFlags.ts
    * Add `isLoadingFeatureFlags` to return values

7. ui/apps/platform/src/hooks/usePermissions.ts
    * Add `isLoadingPermissions` to return values

8. Edit src/reducers/featureFlags.js
    * Add `error` reducer and `getFeatureFlagsError` selector
    * Add `isLoading` reducer and `getIsLoadingFeatureFlags` selector

9. Edit src/reducers/roles.js
    * Add `error` reducer and `getUserRolePermissionsError` selector
    * Add `isLoading` reducer and `getIsLoadingUserRolePermissions` selector

10. Edit src/reducers/roles.test.js
    * Add `error` and `isLoading` to expected initial state

11. Edit src/sagas/authSagas.js
    * Add comment and request action for subsequent requests
    * Add failure action to store error

12. Edit src/sagas/featureFlagSagas.js
    * Add comment and request action for subsequent requests

### Residue

1. Investigate whether `isLoading` is needed for any other data in Redux from sagas: clusters? network graph?
2. Delete obsolete feature flags
3. MainPage.tsx
    * Replace `SearchModal` element with route in Body.tsx: /main/search path and `GlobalSearchPage` element
    * Decide how to render error if request fails for feature flags or permissions

## Checklist

## Testing Performed

1. See same `featureflags` and `mypermissions` requests in Network panel
2. Temporarily comment out `fetchFeatureFlags` in featureFlagSagas.js
    * See expected page crash without `if` statement in `MainPage`
    * See expected endless spinner with `if` statement in `MainPage`
4. Group test
